feat!: rewrite in TypeScript with CacheMap support

aminya opened the pull request:
This pull request rewrites the action in TypeScript and adds support for `cache-map` that gets a string of files that need to be injected as a JSON string. This makes it possible to inject multiple directories in one call and simplifies the usage.

It also makes it possible to run the script outside GitHub Actions (e.g. for S3) or locally using command line arguments. 

BREAKING `cache-source` and `cache-target` are deprecated in favour of `cache-map` that expects a JSON string. 

I have bumped the version to v3.0.0


Fixes #10 
Closes #24
Closes #22  
Allows implementation of #16 


```yaml
      - name: Cache
        uses: actions/cache@v3
        id: cache
        with:
          path: |
            var-cache-apt
            var-lib-apt
          key: cache-${{ hashFiles('.github/workflows/test/Dockerfile') }}

      - name: inject cache into docker
        uses: ./
        with:
          cache-map: |
            {
              "var-cache-apt": "/var/cache/apt",
              "var-lib-apt": "/var/lib/apt"
            }
          skip-extraction: ${{ steps.cache.outputs.cache-hit }}

```

AkihiroSuda commented on action.yml:
```diff
@@ -1,12 +1,9 @@
-    description: "Where the cache is stored in the calling workspace. Default: `cache`"
-  cache-target:
-    default: /root/.cache/go-build
-    description: "Where the cache is stored in the docker container. Default: `/root/.cache/go-build`"
```
Can we continue to support the old form, with a deprecation notice?

aminya replied:
Yeah, I can add a compatibility layer.

AkihiroSuda commented on src/index.ts:
```diff
@@ -4,6 +4,7 @@
 /* ================================================================================================================== *
  * Authors:                                                                                                           *
  *   Unai Martinez-Corral                                                                                             *
```
This guy's code might no longer be present in the current revision?

aminya replied:
Yeah. Should I remove the whole header or just the name?

AkihiroSuda commented on README.md:
```diff
@@ -32,43 +32,39 @@ Action:
+          key: cache-${{ hashFiles('.github/workflows/test/Dockerfile') }}
+
+      - name: inject cache into docker
+        uses: reproducible-containers/buildkit-cache-dance@v3.0.0
```
Could you update the " Releases" section below to explain the changes since v2?

aminya commented:
Recovered the cache-source and cache-target options but with deprecation messages. Also, added an example of running this outside GitHub Actions.

AkihiroSuda commented on dist/index.js:
To avoid having a "binary" blob in the repo, would it be possible to use a plain JavaScript instead of TypeScript?
Otherwise the reproducibility of this "binary" blob would need to be tested in the CI

aminya replied:
This is the official way of doing it based on the template provided by GitHub.
https://github.com/actions/typescript-action
https://github.com/actions/javascript-action

But I can add automatic updates to the dist folder 
https://github.com/actions/setup-node/blob/main/.github/workflows/check-dist.yml

aminya replied:
Fixed in [`8325b31` (#25)](https://github.com/reproducible-containers/buildkit-cache-dance/pull/25/commits/8325b3152229eacef70506be51b175a8f994febd)

AkihiroSuda commented:
https://github.com/reproducible-containers/buildkit-cache-dance/actions/runs/8490107723/job/23260959438?pr=25

```
Post job cleanup.

FROM busybox:1
COPY buildstamp buildstamp
RUN --mount=type=cache,target=/var/cache/apt     mkdir -p /var/dance-cache/     && cp -p -R /var/cache/apt/. /var/dance-cache/ || true


node:internal/process/esm_loader:40
      internalBinding('errors').triggerUncaughtException(
                                ^
tar: Skipping to next header
tar: Exiting with failure status due to previous errors

(Use `node --trace-uncaught ...` to show where the exception was thrown)

Node.js v[2](https://github.com/reproducible-containers/buildkit-cache-dance/actions/runs/8490107723/job/23260959438?pr=25#step:11:2)0.8.1
```

aminya commented:
The error you sent doesn't happen on my fork:
https://github.com/aminya/buildkit-cache-dance/actions

AkihiroSuda commented:
The test passes when some cache is present on GHA, but it still fails after pruning the cache via https://github.com/reproducible-containers/buildkit-cache-dance/actions/caches

AkihiroSuda commented:
Also, please remove the "Merge" commit (use `git rebase`, not `git merge`), and consider squashing the commits

aminya commented:
> The test passes when some cache is present on GHA, but it still fails after pruning the cache via [reproducible-containers/buildkit-cache-dance/actions/caches](https://github.com/reproducible-containers/buildkit-cache-dance/actions/caches)

Okay, found the bug. The stdout was not being captured properly.

aminya commented:
Simplified and fixed the docker to tar data piping issue. It now works on my fork with cache pruned as well:
https://github.com/aminya/buildkit-cache-dance/actions/runs/8494967106/job/23270766398

AkihiroSuda reviewed (approved):
Thanks

strophy commented:
Thanks @aminya this is a really significant improvement!

aminya commented:
You're welcome!
